tests/cpptests/test_cpp_query.cpp

```diff
@@ -447,15 +447,17 @@ TEST_F(QueryTest, testParser_v3) {
   // Tag queries
   assertValidQuery("@tags:{foo}", ctx);
   assertValidQuery("@tags:{foo|bar baz|boo}", ctx);
-  assertValidQuery("@tags:{foo|bar\\ baz|boo}", ctx);
+  assertInvalidQuery("@tags:{foo|bar\\ baz|boo}", ctx);
```

Comment:
Why is this invalid?

Reply:
`@tags:{foo|bar\\ baz|boo}"`  should be invalid because only the `tag_invalid_punct = (rb | star | escape | '$')` should be escaped in a tag, and the expression is escaping space ` `.

But there is an error because the parser is accepting `TERM` (`escaped_terms` in the lexer) as part of the `tag_list`, using the rules:
```sh
term(A) ::= TERM(B)
param_term_case(A) ::= term(B)
tag_list(A) ::= param_term_case(B) . [TAGLIST]
```
In the example, we also have pipes, thus the parser is trying to handle multiple `TERM` and fails.

We can see an additional problem with this case:
```sh
127.0.0.1:6379> FT.explaincli idx "@t:{foo bar\\ baz}" DIALECT 5
1) TAG:@t {
2)   INTERSECT {
3)     foo
4)     bar baz
5)   }
6) }
```

A similar issue is shown with the prefix/suffix/infix, this should be an invalid query:
`assertValidQuery("@tags:{foo\\-*}", ctx);`

We need to change the parser to reject `TERM`s inside the `{}`, for the cases of exact matching, and prefix/suffix/infix.

Reply:
This PR [538002d](https://github.com/RediSearch/RediSearch/pull/4433/commits/538002dcc85754d6acf359eecdc5625ae40c5235) modifies the parser to fix the issue.
The rules for the `tag_list` were replaced by `single_tag` rules, and now the rules does not accept terms:
```sh
single_tag(A) ::= ATTRIBUTE(B)
single_tag(A) ::= UNESCAPED_TAG(B)
single_tag(A) ::= affix_tag(B)
single_tag(A) ::= verbatim(B)

affix_tag(A) ::= PREFIX_TAG(B)
affix_tag(A) ::= SUFFIX_TAG(B)
affix_tag(A) ::= CONTAINS_TAG(B)
```

Due to the previous change, we need to fix the lexer to support parameters as part of prefix/suffix/infix:
```sh
contains_tag = lb (star.single_tag.star | star.attr.star) rb $1;
prefix_tag = lb (single_tag.star | attr.star) rb $1;
suffix_tag = lb (star.single_tag | star.attr) rb $1;
```

New test to validate the use of valid tag characters: 
[test_tags.py:testDialect5ValidChars()](https://github.com/RediSearch/RediSearch/pull/4433/commits/538002dcc85754d6acf359eecdc5625ae40c5235#diff-b71b7402df7c4b83c0e5690ef9c4d133ddb694ca43d2371f648bcfe5f23267a2R800)

Reply:
Justification: Similar to the JSON format, where an escape can only precede a specific list of characters. Escaping other characters is considered a syntax error.